contrib/node/src/python/pants/contrib/node/tasks/node_resolve.py

```diff
@@ -66,23 +72,58 @@ def _can_resolve_target(self, target):
     """
     return self.is_node_package(target) and self._resolver_for_target(target) != None
 
+  def _do_local_invalidation_check(self,
+                                   targets,
+                                   invalidate_dependents=True,
+                                   topological_order=True):
+    """Returns an invalidation_check with all targets invalidated.
+    """
+    # Generate a cache_key that never matches to ensure work is always done.
+    cache_key_generator = UncacheableCacheKeyGenerator()
+    cache_manager = InvalidationCacheManager(self.workdir,
+                                             cache_key_generator,
+                                             self._build_invalidator,
+                                             invalidate_dependents,
+                                             fingerprint_strategy=None,
+                                             invalidation_report=self.context.invalidation_report,
+                                             task_name=self._task_name,
+                                             task_version=self.implementation_version_str())
+
+    return cache_manager.check(targets, topological_order=topological_order)
+
   def execute(self):
     targets = self.context.targets(predicate=self._can_resolve_target)
     if not targets:
       return
-
-    node_paths = self.context.products.get_data(NodePaths, init_func=NodePaths)
-
-    # We must have copied local sources into place and have node_modules directories in place for
-    # internal dependencies before installing dependees, so `topological_order=True` is critical.
-    with self.invalidated(targets,
-                          topological_order=True,
-                          invalidate_dependents=True) as invalidation_check:
-
-      with self.context.new_workunit(name='install', labels=[WorkUnitLabel.MULTITOOL]):
+    if self.context.products.is_required_data(NodePaths):
+      node_paths = self.context.products.get_data(NodePaths, init_func=NodePaths)
+      # We must have copied local sources into place and have node_modules directories in place for
+      # internal dependencies before installing dependees, so `topological_order=True` is critical.
+      with self.invalidated(targets,
+                            topological_order=True,
+                            invalidate_dependents=True) as invalidation_check:
+        with self.context.new_workunit(name='install', labels=[WorkUnitLabel.MULTITOOL]):
+          for vt in invalidation_check.all_vts:
+            target = vt.target
+            if not vt.valid:
+              resolver_for_target_type = self._resolver_for_target(target).global_instance()
+              resolver_for_target_type.resolve_target(self, target, vt.results_dir, node_paths)
+            node_paths.resolved(target, vt.results_dir)
+    if self.context.products.is_required_data(NodePathsLocal):
```

Comment:
Are there any cases in which this should represent a symlink into `.pants.d` rather than always double installing things?

Reply:
By exposing two products and treating them both as optional, we effectively removed the "double installation" problem.

Reply:
Except when they've both been requested?

It's not critical, it's just unfortunate that `./pants node-install` can never (re)use the work done by the private copy.